pkg/controller/networkpolicy/validate.go

```diff
@@ -20,13 +20,46 @@ import (
 	"strconv"
 
 	admv1 "k8s.io/api/admission/v1"
+	authenticationv1 "k8s.io/api/authentication/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/klog"
 
 	secv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/security/v1alpha1"
 )
 
+// validator interface introduces the set of functions that must be implemented
+// by any resource validator.
+type validator interface {
+	// createValidate is the interface which must be satisfied for resource
+	// CREATE events.
+	createValidate(curObj interface{}, userInfo authenticationv1.UserInfo, v *NetworkPolicyValidator) (string, bool)
+	// updateValidate is the interface which must be satisfied for resource
+	// UPDATE events.
+	updateValidate(curObj, oldObj interface{}, userInfo authenticationv1.UserInfo, v *NetworkPolicyValidator) (string, bool)
+	// deleteValidate is the interface which must be satisfied for resource
+	// DELETE events.
+	deleteValidate(oldObj interface{}, userInfo authenticationv1.UserInfo, v *NetworkPolicyValidator) (string, bool)
```

Comment:
wonder if `v *NetworkPolicyValidator` should be a part of the interface here?

It seems that all the state is encapsulated in `networkPolicyController` and that's what `antreaPolicyValidator` and `tierValidator` need to have access to. If I think of the ability to register extra validators (probably defined outside of this package), which is the point of this PR, then there isn't much point in being able to access `NetworkPolicyValidator` in `createValidate`, etc., since it doesn't have useful public methods (besides `Validate` but of course a `validator` is not going to call that).

Reply:
updated